wp-includes/mysql/class-mysql-lexer.php

```diff
@@ -0,0 +1,2752 @@
+		}
+		$this->type = self::INT_NUMBER;
+	}
+
+	protected function decimal_number() {
+		$this->consume(); // Consume the '.'.
+		while ( $this->is_digit( $this->c ) ) {
+			$this->consume();
+		}
+		$this->type = self::DECIMAL_NUMBER;
+	}
+
+	protected function line_comment() {
+		while ( null !== $this->c && "\n" !== $this->c && "\r" !== $this->c ) {
+			$this->consume();
+		}
+		$this->type    = self::COMMENT;
+		$this->channel = self::CHANNEL_HIDDEN;
+	}
+
+	protected function block_comment() {
+		$this->consume(); // Consume the '/'.
+		$this->consume(); // Consume the '*'.
+		$this->skip_comment_content();
+		$this->type    = self::COMMENT;
+		$this->channel = self::CHANNEL_HIDDEN;
+	}
+
+	protected function mysql_comment() {
+		// MySQL-specific comment in one of the following forms:
+		//   1. /*! ... */      - The content is treated as regular SQL code.
+		//   2. /*!12345 ... */ - The content is treated as SQL code when "server version >= 12345".
+		$this->consume(); // Consume the '/'.
+		$this->consume(); // Consume the '*'.
+		$this->consume(); // Consume the '!'.
+
+		// Check if the next 5 characters are digits.
+		$is_version_comment = true;
+		for ( $i = 0; $i < 5; $i++ ) {
+			if ( ! $this->is_digit( $this->c ) ) {
+				$is_version_comment = false;
+				break;
+			}
+			$this->consume(); // Consume the digit.
+		}
+
+		// For version comments, extract the version number.
+		$version = $is_version_comment
+			? (int) substr( $this->get_text(), 3 ) // Strip the '/*!' prefix.
+			: 0;
+
+		if ( $this->server_version < $version ) {
+			// When version not satisfied. Treat the content as a regular comment.
+			$this->skip_comment_content();
+			$this->type = self::COMMENT;
+		} else {
+			// Version satisfied or not specified. Treat the content as SQL code.
+			$this->in_version_comment = true;
+			$this->type               = self::MYSQL_COMMENT_START;
+		}
+		$this->channel = self::CHANNEL_HIDDEN;
+	}
+
+	protected function skip_comment_content() {
+		while ( null !== $this->c ) {
+			if ( '*' === $this->c && '/' === $this->n ) {
+				$this->consume(); // Consume the '*'.
+				$this->consume(); // Consume the '/'.
+				break;
+			}
+			$this->consume();
+		}
+	}
+
+	// Helper functions -----------------------------------------------------------------------------------------------------
+
+	private function is_whitespace( $char ) {
+		return isset( self::WHITESPACES[ $char ] );
+	}
+
+	private function is_digit( $char ) {
+		return $char >= '0' && $char <= '9';
+	}
+
+	private function determine_numeric_type( $text ) {
+		if ( preg_match( '/^0[xX][0-9a-fA-F]+$/', $text ) ) {
+			return self::HEX_NUMBER;
+		} elseif ( preg_match( '/^0[bB][01]+$/', $text ) ) {
+			return self::BIN_NUMBER;
+		} elseif ( preg_match( '/^\d+$/', $text ) ) {
+			if ( PHP_INT_MAX >= (int) $text ) {
+				return self::INT_NUMBER;
+			}
+			return self::LONG_NUMBER;
+		}
+		return self::INVALID_INPUT;
+	}
+}
+
+function tokenize_query( $sql ) {
```

Comment:
How about `WP_MySQL_Lexer::tokenize()`? It's less opportunity for function name conflict and it's also explicit that we're talking about MySQL queries.

Reply:
@adamziel The lexer has some optional arguments: `__construct( string $input, int $server_version = 80038, int $sql_modes = 0 )`. I wonder if it makes sense to make a `WP_MySQL_Lexer::tokenize()` static method, which would probably need to mirror them, or make it an instance one:

```php
$lexer = new WP_MySQL_Lexer($sql, 80038, WP_MySQL_Lexer::SQL_MODE_ANSI_QUOTES);
$lexer->tokenize(); // get all tokens
// or
$lexer->get_next_token(); // "stream" them to the parser one by one
```

Or we could `yield` them and then `tokenize` could handle both. I guess it also depends on what's the precedent in WP and how we want the parser API to look like (discussed in https://github.com/WordPress/sqlite-database-integration/pull/157#discussion_r1827588161). The [WP_HTML_Tag_Processor](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/html-api/class-wp-html-tag-processor.php#L926) is not exactly a lexer and `next_token`/`next_tag` don't necessarily have the same meaning.

Reply:
Good points!

Streaming tokens be useful in the generalized parser, but in the MySQL one I don't see a need for it. It would complicate things for no benefit - all the SQL queries must fit into memory or we won't be able to pass them to mysql_query().

At the same time, it would be useful to raw read bytes from a SQL export and get one SQL query after another. It doesn't have to be a part of the tokenizer, but the tokenizer would have to understand the difference between "we have incomplete data" vs "it's a syntax error".

Baking this into the scanner sounds like a substantial project, so perhaps we could fake it somehow. Or let go of the idea for now and focus on wrapping this up and moving on to the driver.

Reply:
Aha, gotcha! I like what you proposed for the API. Perhaps we could even do away with:

* `remaining_tokens(): Token[]|bool`
* `next_token(): Token|bool`

`false` would mean either that the tokenization has finished, errored out, or, in the future, that we're in a paused state waiting for more data. What do you think?

Reply:
> There is also the question whether the lexer should have a new instance for each query or you'd have it like a service, where a single instance would process multiple queries.

One lexer per query might not work if we don't know upfront where one query starts and the other query ends. Suppose we have the following SQL file:

```sql
SELECT '<?php echo "a";'; SELECT * FROM wp_posts;
```

Now imagine we do `fread()` in loop and get the following chunks:

```php
$chunk1 = 'SELECT \'<?php echo "a";'
$chunk2 = '\'; SELECT * FROM wp_posts;';
```

There's nothing obvious telling us where to split. We need to tokenize until we get to the semicolon token or EOF and only then we know we're at the end of the query. Even worse, the first chunk is a syntax error on its own. Only after getting the second chunk we have a valid query to parse.

I've had the same problem when working on `WP_XML_Processor` and went for an explicit `$this->expecting_more_input` flag. When there's not enough input but we're still expecting more, incomplete tokens cause a state transition to `STATE_INCOMPLETE_INPUT`. Then, calling `$xml->append_bytes( $chunk );` causes a state transition to `STATE_READY`. However, if we see incomplete input once `$this->expecting_more_input === false`, it's a syntax error and we finish the parsing.

We could do a similar thing here if it's relatively easy. If it's difficult, let's leave that problem for another day.

Reply:
@adamziel Ah, thanks for sharing, that's a great insight and really good to know how it was already handled in a similar case!

> One lexer per query might not work when we don't knowupfront where one query ends and the other query starts.

This option didn't occur to me and thinking about it, I think what I meant could work if we say "one lexer per input". In the case of the MySQL parser, the nice thing is that multi-queries are defined by the grammar. The following already works:

```php
$tokens     = WP_MySQL_Lexer::tokenize( 'SELECT 1; SELECT 2; SELECT 3;' );
$parser     = new WP_MySQL_Parser( $grammar, $tokens );
$parse_tree = $parser->parse();
```

Although I just noticed that this works too, and the last invalid query is ignored:
```php
'SELECT 1; SELECT 2; some invalid input'
```
So we may need to fix this case. Probably won't be too hard.

Anyway, I guess "one lexer per input" can work, unless we would prefer a service-like API:
```php
$lexer = new WP_MySQL_Lexer( $config );
$lexer->set_input( ... );
$lexer->next_token();
$lexer->remaining_tokens();

// reuse the same instance
$lexer->set_input( ... );
$lexer->next_token();
$lexer->remaining_tokens();

// etc.
```

I guess there won't be a large performance impact on new lexer & parser per input, but we can benchmark that to be sure. In any way, what do you think about these APIs — new lexer/parser instances per input vs. service-like interfaces? Do you have any preferences and do we have any more inspiration in the WP codebase?

Maybe in case of a parser, the expectation could be that you can use it over and over again:

```php
$parser = new WP_MySQL_Parser( ... );
$parser->parse( $query1 );
$parser->parse( $query2 );
// etc.
```

I'm not sure.

As for `next_token()` and `remaining_tokens()`, that sounds great to me!

Reply:
Something like `$tokens     = WP_MySQL_Lexer::tokenize( 'SELECT 1; SELECT 2; SELECT 3;' );` is great for scenarios where multiple mysql queries are crammed into a single function call – we need that to support those.

Where it falls short is processing a 50GB SQL export that we can't load into memory. For this, we'd need the ability to append SQL bytes and clean up previous bytes as we go.

API-wise, yeah maybe we don't need a static call and we can just create a new WP_MySQL_Lexer instance for now. One lexer per _SQL stream_ sounds great. One parser per _SQL query_ also sounds great. Let's start with a restrictive API that requires multiple WP_MySQL_Parser instances and doesn't allow multiple `parse()` calls and see if that's a problem at all when using it in the database driver. This isn't a big architectural difference so we can always backtrack.